scripts/build_osm_network.py

```diff
@@ -0,0 +1,1051 @@
+    tolerance (float): Tolerance for coordinate deviation.
+
+    Returns:
+    bool: True if the points are almost equal, False otherwise.
+    """
+    return abs(point1.x - point2.x) < tolerance and abs(point1.y - point2.y) < tolerance
+
+
+def merge_linestrings(gdf):
+    """
+    Merge LineStrings in a GeoDataFrame wherever the endpoints match.
+
+    Parameters:
+    gdf (GeoDataFrame): A GeoDataFrame containing LineString geometries.
+
+    Returns:
+    GeoDataFrame: A GeoDataFrame with merged LineString geometries.
+    """
+    gdf = gdf.copy()
+    if len(gdf) == 1:
+        return gdf
+
+    lines = list(gdf.geometry)
+    merged_lines = []
+    while lines:
+        line = lines.pop(0)
+        merged_line = line
+        i = 0
+        while i < len(lines):
+            if are_almost_equal(
+                Point(merged_line.coords[-1]), Point(lines[i].coords[0])
+            ):
+                merged_line = LineString(
+                    list(merged_line.coords) + list(lines.pop(i).coords[1:])
+                )
+                i = 0  # Restart the scan after merging
+            elif are_almost_equal(
+                Point(merged_line.coords[0]), Point(lines[i].coords[-1])
+            ):
+                merged_line = LineString(
+                    list(lines.pop(i).coords)[:-1] + list(merged_line.coords)
+                )
+                i = 0  # Restart the scan after merging
+            elif are_almost_equal(
+                Point(merged_line.coords[-1]), Point(lines[i].coords[-1])
+            ):
+                merged_line = LineString(
+                    list(merged_line.coords) + list(lines.pop(i).coords[::-1])[1:]
+                )
+                i = 0  # Restart the scan after merging
+            elif are_almost_equal(
+                Point(merged_line.coords[0]), Point(lines[i].coords[0])
+            ):
+                merged_line = LineString(
+                    list(lines.pop(i).coords[::-1])[:-1] + list(merged_line.coords)
+                )
+                i = 0  # Restart the scan after merging
+            else:
+                i += 1
+        merged_lines.append(merged_line)
+        no_coordinates = [len(merged_lines[i].coords) for i in range(len(merged_lines))]
+        max_index = np.argmax(no_coordinates)
+        merged_lines = [merged_lines[max_index]]
+
+    return gpd.GeoDataFrame(geometry=merged_lines, crs=gdf.crs)
+
+
+if __name__ == "__main__":
+    # Detect running outside of snakemake and mock snakemake for testing
+    if "snakemake" not in globals():
+        from _helpers import mock_snakemake
+
+        snakemake = mock_snakemake("build_osm_network")
+
+    configure_logging(snakemake)
+    set_scenario_config(snakemake)
+
+    # load default crs
+    geo_crs = "EPSG:4326"
+    distance_crs = "EPSG:3035"
+
+    countries = snakemake.config["countries"]
+
+    with memory_logger(
+        filename=getattr(snakemake.log, "memory", None), interval=30.0
+    ) as mem:
```

Comment:
Is the memory logger critical here? Could remove. If not, is 30s-interval adequate?

Reply:
Not needed anymore, only used for debugging purposes during development.